### ISSUE TITLE
GetResponse timeout on second and subsequent backup of large dataset to OneDrive in 2.0.0.80

### ISSUE THREAD

**Issue by davidwbryant** [duplicati/duplicati#1299]
Duplicati.Library.Main.BackendManager.List() consistently fails for the OneDrive backend with a timeout when the size of the backup is quite large - 165GB in my case.

Platform is Windows 7 Ultimate.

Increasing the timeout results in a failure due to an exception to say that the underlying connection has been closed with an unexpected error.

There is a fair bit of discussion on the web about this exception, but none of the suggested solutions seems to make any difference.

Where the exception occurs in Duplicati/Library/Backend/OneDrive/OneDrive.cs a request is made with the OneDrive API for FOLDER_ID/files, however in this case there are over 6000 files in the folder for the backup in question.

The Microsoft OneDrive API allows the files request to be broken down as:
FOLDER_ID/files?limit=x&offset=y.

Implementing this for the List() method solves the problem for my backup, although I haven't been able to test yet with other similarly sized backups due to the amount of data that would need to be uploaded.

Code changes look like this - I'm sure they could be better, I'm not a c# developer

```
+       private const string FOLDER_TEMPLATE1 = "{0}/files?limit={1}&offset={2}";

         public List<IFileEntry> List()

         {

-            var url = string.Format("{0}/{1}?access_token={2}", WLID_SERVER, string.Format(FOLDER_TEMPLATE, FolderID), Library.Utility.Uri.UrlEncode(AccessToken));

-

-            var res = GetJSONData<WLID_DataItem>(url);

+            // When the backup becomes large, the list operation may fail by causing a timeout on get response, or if 

+            // the timeout is extended by throwing an exception - The underlying connection was closed: An unexpected error occurred on a receive

+            // Therefore limit the number of files requested at a time here (see https://msdn.microsoft.com/en-gb/library/hh826531.aspx)

+            int filecount = 100;

+            int offset = 0;

+            

+            Logging.Log.WriteMessage("OneDrive->List ", Logging.LogMessageType.Error);

+            

             var files = new List<IFileEntry>();

 

             m_fileidCache.Clear();

 

-            if (res != null && res.data != null)

-                foreach (var r in res.data)

+            while (true)

+            {

+                var outMsg = string.Format("filecount {0} offset {1}", filecount, offset);

+                Logging.Log.WriteMessage(outMsg, Logging.LogMessageType.Error);

+                var url = string.Format("{0}/{1}&access_token={2}", WLID_SERVER, string.Format(FOLDER_TEMPLATE1, FolderID, filecount, offset), Library.Utility.Uri.UrlEncode(AccessToken));

+                var res = GetJSONData<WLID_DataItem>(url);

+                int count = 0;

+

+                if (res != null && res.data != null)

                 {

-                    m_fileidCache.Add(r.name, r.id);

+                    foreach (var r in res.data)

+                    {

+                        m_fileidCache.Add(r.name, r.id);

 

-                    var fe = new FileEntry(r.name, r.size, r.updated_time, r.updated_time);

-                    fe.IsFolder = string.Equals(r.type, "folder", StringComparison.InvariantCultureIgnoreCase);

-                    files.Add(fe);

+                        var fe = new FileEntry(r.name, r.size, r.updated_time, r.updated_time);

+                        fe.IsFolder = string.Equals(r.type, "folder", StringComparison.InvariantCultureIgnoreCase);

+                        files.Add(fe);

+                        ++count;

+                    }

+                    if (count == 0)

+                        break;

+                    else

+                        offset += filecount;

                 }

-

+                else

+                {

+                    break;

+                }

+            }

             return files;

         }
```

**Comment by kenkendk** [comment]
Great work tracking down the problem.

I will see if I can test this.

**Comment by cheese528** [comment]
Hi,
I seem to have hit this same problem again with a larger dataset.  I have 521GB in source data that is taking up 446GB in OneDrive.  I had added files to the backup incrementally using folder exclusions, and up to around 400GB+ of source data, I did not have any issues adding the files.  One the very last backup (the 7th version) , I had these time-out errors:

`System.Net.WebException: GetResponse timed out ---> System.Net.WebException: The request was aborted: The request was canceled. at System.Net.HttpWebRequest.EndGetResponse(IAsyncResult asyncResult) at Duplicati.Library.Utility.AsyncHttpRequest.AsyncWrapper.OnAsync(IAsyncResult r) --- End of inner exception stack trace --- at Duplicati.Library.Main.BackendManager.List() at Duplicati.Library.Main.Operation.FilelistProcessor.RemoteListAnalysis(BackendManager backend, Options options, LocalDatabase database, IBackendWriter log) at Duplicati.Library.Main.Operation.FilelistProcessor.VerifyRemoteList(BackendManager backend, Options options, LocalDatabase database, IBackendWriter log) at Duplicati.Library.Main.Operation.BackupHandler.PreBackupVerify(BackendManager backend) at Duplicati.Library.Main.Operation.BackupHandler.Run(String[] sources, IFilter filter) at Duplicati.Library.Main.Controller.<Backup>c__AnonStorey0.<>m__0(BackupResults result) at Duplicati.Library.Main.Controller.RunAction[T](T result, String[]& paths, IFilter& filter, Action``1 method) at Duplicati.Library.Main.Controller.Backup(String[] inputsources, IFilter filter) at Duplicati.Server.Runner.Run(IRunnerData data, Boolean fromQueue)`

Currently running 2.0.1.32_canary_2016-11-12 with the default 50MB volume size.   Am I hitting the limits of the OneDrive API, or is this something related to the timeout values?  Hopefully this can be fixed so that I don't need to abandon my backup.

At this point, I can't even run "Test connection" on this folder in OneDrive.  Setting it to another path on the server is still fine though so the AuthID still works as far as I can tell.
